HBASE-29267 Support shaded clients in Exception deserialization by prefixing shaded package in IPCUtil

mwkang opened the pull request:
[HBASE-29267](https://issues.apache.org/jira/browse/HBASE-29267)

mwkang committed:
HBASE-29267 Support shaded clients in Exception deserialization by prefixing shaded package in IPCUtil

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 36s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 51s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 53s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 19s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |   0m 51s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 51s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 24s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 20s | [/results-checkstyle-hbase-client.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/artifact/yetus-general-check/output/results-checkstyle-hbase-client.txt) |  hbase-client: The patch generated 1 new + 1 unchanged - 0 fixed = 2 total (was 1)  |
| +1 :green_heart: |  spotbugs  |   0m 59s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  13m 34s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  spotless  |   0m 58s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 13s |  |  The patch does not generate ASF License warnings.  |
|  |   |  36m 37s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6917 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless |
| uname | Linux 7db3c4edfe8f 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 7c847808a29f1c935628fe04314af66e97533d6d |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 82 (vs. ulimit of 30000) |
| modules | C: hbase-client U: hbase-client |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   6m 20s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  3s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   4m 22s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 26s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 27s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   7m 48s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 52s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 35s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 35s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 26s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   8m  6s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |   1m 59s | [/patch-unit-hbase-client.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/artifact/yetus-jdk17-hadoop3-check/output/patch-unit-hbase-client.txt) |  hbase-client in the patch failed.  |
|  |   |  38m 28s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6917 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 69a78ed48eaf 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / 7c847808a29f1c935628fe04314af66e97533d6d |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/testReport/ |
| Max. process+thread count | 280 (vs. ulimit of 30000) |
| modules | C: hbase-client U: hbase-client |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/1/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 32s |  |  Docker mode activated.  |
| -0 :warning: |  yetus  |   0m  2s |  |  Unprocessed flag(s): --brief-report-file --spotbugs-strict-precheck --author-ignore-list --blanks-eol-ignore-file --blanks-tabs-ignore-file --quick-hadoopcheck  |
|||| _ Prechecks _ |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 13s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 20s |  |  master passed  |
| +1 :green_heart: |  javadoc  |   0m 17s |  |  master passed  |
| +1 :green_heart: |  shadedjars  |   5m 56s |  |  branch has no errors when building our shaded downstream artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m  4s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  shadedjars  |   5m 52s |  |  patch has no errors when building our shaded downstream artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 35s |  |  hbase-client in the patch passed.  |
|  |   |  22m 38s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/artifact/yetus-jdk17-hadoop3-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6917 |
| Optional Tests | javac javadoc unit compile shadedjars |
| uname | Linux 097e06808130 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / ae2040476d29dfaa0fe1795dc21ff948ddc0b69f |
| Default Java | Eclipse Adoptium-17.0.11+9 |
|  Test Results | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/testReport/ |
| Max. process+thread count | 325 (vs. ulimit of 30000) |
| modules | C: hbase-client U: hbase-client |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/console |
| versions | git=2.34.1 maven=3.9.8 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

Apache-HBase commented:
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   6m 27s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  |  Patch does not have any anti-patterns.  |
|||| _ master Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   3m 30s |  |  master passed  |
| +1 :green_heart: |  compile  |   0m 48s |  |  master passed  |
| +1 :green_heart: |  checkstyle  |   0m 16s |  |  master passed  |
| +1 :green_heart: |  spotbugs  |   0m 43s |  |  master passed  |
| +1 :green_heart: |  spotless  |   0m 56s |  |  branch has no errors when running spotless:check.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   6m 30s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 49s |  |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 49s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 15s | [/results-checkstyle-hbase-client.txt](https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/artifact/yetus-general-check/output/results-checkstyle-hbase-client.txt) |  hbase-client: The patch generated 1 new + 1 unchanged - 0 fixed = 2 total (was 1)  |
| +1 :green_heart: |  spotbugs  |   0m 55s |  |  the patch passed  |
| +1 :green_heart: |  hadoopcheck  |  21m  4s |  |  Patch does not cause any errors with Hadoop 3.3.6 3.4.0.  |
| +1 :green_heart: |  spotless  |   0m 52s |  |  patch has no errors when running spotless:check.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  asflicense  |   0m 11s |  |  The patch does not generate ASF License warnings.  |
|  |   |  56m 54s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/6917 |
| Optional Tests | dupname asflicense javac spotbugs checkstyle codespell detsecrets compile hadoopcheck hbaseanti spotless |
| uname | Linux 28dfe1dab58b 5.4.0-1103-aws #111~18.04.1-Ubuntu SMP Tue May 23 20:04:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/hbase-personality.sh |
| git revision | master / ae2040476d29dfaa0fe1795dc21ff948ddc0b69f |
| Default Java | Eclipse Adoptium-17.0.11+9 |
| Max. process+thread count | 83 (vs. ulimit of 30000) |
| modules | C: hbase-client U: hbase-client |
| Console output | https://ci-hbase.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-6917/2/console |
| versions | git=2.34.1 maven=3.9.8 spotbugs=4.7.3 |
| Powered by | Apache Yetus 0.15.0 https://yetus.apache.org |


This message was automatically generated.

junegunn reviewed (commented):
Thanks.

I've noticed that this patch can be incompatible with some shading configurations.

For example, here's one configuration we evaluated for using both HBase 1 and 2 clients in the same project:

```
tasks.shadowJar {
    relocate("org.apache.hadoop.hbase", "shaded.hbase1") {
        exclude("org/apache/hadoop/hbase/codec/**")

        // Exclude exceptions except protobuf and shaded packages
        exclude("%regex[org/apache/hadoop/hbase/((?!protobuf|shaded)[^/]+/.*)?[^/]+Exception]")
    }
    archiveClassifier.set("")
}
```

In this setup, we are replacing `org.apache.hadoop.hbase` with `shaded.hbase1`, rather than simply prepending a custom prefix, a strategy this patch currently assumes. As a result, we get:

```
ipc.ShadedPrefixUtil: Cannot find shaded pattern for shaded.hbase1.ipc
```

Also, note that we intentionally avoided shading `*Exception` classes to address the original issue of not recognizing exception names from the server. So modifying the class name would cause `ClassNotFoundException`.

mwkang commented:
Thank you for your review.
Sorry for the late reply.
I will check the points you mentioned.

mwkang commented:
Thank you again for the detailed feedback. I've updated the patch to address both concerns you raised.

**1. Support for full-replacement shading**

The previous detection logic searched for .org.apache.hadoop.hbase. as a substring, which only worked
for prefix-prepend configurations (e.g. a.b.c.org.apache.hadoop.hbase.ipc). It silently fell back to
no-op for full-replacement configurations like yours (shaded.hbase1.ipc).

The new approach derives the shaded base by stripping the known .ipc suffix from the current package
name:
- a.b.c.org.apache.hadoop.hbase.ipc → shadedBase = a.b.c.org.apache.hadoop.hbase
- shaded.hbase1.ipc → shadedBase = shaded.hbase1

This covers both strategies uniformly.

**2. Fallback for selectively excluded exception classes**

The new resolveShading() method attempts to load the transformed class name via Class.forName. If the
shaded class is not found (e.g. because *Exception classes were excluded from relocation), it falls
back to the original name. The result is memoized in a ConcurrentHashMap so Class.forName is invoked
at most once per distinct class name.

junegunn commented on hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/ShadedPrefixUtil.java:
```diff
@@ -0,0 +1,129 @@
+
+  /**
+   * Cache from original class name to its resolved (shaded or original) equivalent. Class.forName
+   * for a missing class is expensive as it searches all classloaders, so we memoize the result to
```
> all classloaders

We pass a ClassLoader to `Class.forName`, so I'd tighten up the wording.

mwkang replied:
Thanks — reworded to "walks the supplied classloader's parent delegation chain" in 6f5a5c0 to match the actual call.

junegunn replied:
Sorry to nitpick, but maybe we can just drop the second sentence. The `Class.forName` cost detail feels a little out of place and not really needed here. `Class.forName` is already mentioned in the comment near the call site.

mwkang replied:
Good point, dropped the second sentence in 01cff7fb7b. Thanks!

mwkang committed:
HBASE-29267 Address review feedback on ShadedPrefixUtil

- Mark ShadedPrefixUtil as final to satisfy Checkstyle FinalClass rule
  (class only has a private constructor).
- Tighten the Class.forName cost comment to reflect that a specific
  classloader is supplied (its parent delegation chain), not "all
  classloaders".

mwkang commented on hbase-client/src/main/java/org/apache/hadoop/hbase/ipc/ShadedPrefixUtil.java:
```diff
@@ -0,0 +1,129 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ShadedPrefixUtil {
```
Fixed in 6f5a5c0 — marked the class `final`.

mwkang committed:
HBASE-29267 Address review feedback on ShadedPrefixUtil

Drop Class.forName cost detail from cache field comment per review.
